test/Constraints/async.swift

```diff
@@ -7,3 +7,59 @@ func testNonConversions() async {
   let _: () -> Void = doAsynchronously // expected-error{{cannot convert value of type '() async -> ()' to specified type '() -> Void'}}
   let _: () async -> Void = doSynchronously // expected-error{{cannot convert value of type '() -> ()' to specified type '() async -> Void'}}
 }
+
+// Overloading
+func overloaded() -> String { "synchronous" }
+func overloaded() async -> Double { 3.14159 }
+
+func testOverloadedSync() {
+  let _ = overloaded()
```

Comment:
Maybe add a test which makes sure that type-checker did prefer sync in synchronous context and `async` in asynchronous? This test only make sure that there is no ambiguity but it doesn't verify what got picked...

Reply:
It's relying on `await` checking to do that, which has a pile of tests.

Reply:
Sounds good!